File: artifacts/api-server/src/routes/uploads.ts
import { Router, type IRouter } from "express";
import path from "node:path";
import fs from "node:fs/promises";
import crypto from "node:crypto";
import { requireAdmin } from "./admin";

export const UPLOADS_DIR = path.resolve(process.cwd(), "uploads");

const router: IRouter = Router();

const MAX_BYTES = 12 * 1024 * 1024;

router.post("/admin/upload", requireAdmin, async (req, res) => {
  const filename = typeof req.body?.filename === "string" ? req.body.filename : "image";
  const dataUrl = typeof req.body?.dataUrl === "string" ? req.body.dataUrl : "";

  const m = dataUrl.match(/^data:([a-zA-Z0-9.+/-]+);base64,(.+)$/);
  if (!m) {
    res.status(400).json({ message: "Invalid data URL" });
    return;
  }

  const mime = m[1].toLowerCase();
  if (!mime.startsWith("image/")) {
    res.status(400).json({ message: "Only image uploads are supported" });
    return;
  }

  const buf = Buffer.from(m[2], "base64");
  if (buf.length === 0) {
    res.status(400).json({ message: "Empty file" });
    return;
  }
  if (buf.length > MAX_BYTES) {
    res.status(413).json({ message: "Image too large (max 12 MB)" });
    return;
  }

  const fallbackExt = (mime.split("/")[1] || "bin").replace(/[^a-z0-9]/gi, "").slice(0, 6) || "img";
  const safe = filename.replace(/[^a-z0-9._-]/gi, "_").slice(0, 60) || "image";
  const hasExt = /\.[a-z0-9]{2,5}$/i.test(safe);
  const stamp = crypto.randomBytes(8).toString("hex");
  const finalName = hasExt ? `${stamp}-${safe}` : `${stamp}-${safe}.${fallbackExt}`;

  await fs.mkdir(UPLOADS_DIR, { recursive: true });
  await fs.writeFile(path.join(UPLOADS_DIR, finalName), buf);

  res.json({ url: `/api/uploads/${finalName}`, bytes: buf.length });
});

export default router;

File: artifacts/api-server/app/routers/uploads.py
import base64
import os
import re
import secrets
from pathlib import Path
from typing import Any

from fastapi import APIRouter, Depends, Request, status
from fastapi.responses import JSONResponse

from ..auth import require_admin

router = APIRouter()

UPLOADS_DIR = Path(os.getcwd()) / "uploads"

MAX_BYTES = 12 * 1024 * 1024  # 12 MB
DATA_URL_RE = re.compile(r"^data:([a-zA-Z0-9.+/-]+);base64,(.+)$", re.DOTALL)
SAFE_FILENAME_RE = re.compile(r"[^a-z0-9._-]", re.IGNORECASE)
HAS_EXT_RE = re.compile(r"\.[a-z0-9]{2,5}$", re.IGNORECASE)
EXT_CLEAN_RE = re.compile(r"[^a-z0-9]", re.IGNORECASE)


@router.post("/admin/upload", dependencies=[Depends(require_admin)])
async def upload(request: Request):
    try:
        body: Any = await request.json()
    except Exception:
        body = None

    if not isinstance(body, dict):
        body = {}

    raw_filename = body.get("filename")
    filename = raw_filename if isinstance(raw_filename, str) else "image"

    raw_data_url = body.get("dataUrl")
    data_url = raw_data_url if isinstance(raw_data_url, str) else ""

    match = DATA_URL_RE.match(data_url)
    if not match:
        return JSONResponse(
            status_code=status.HTTP_400_BAD_REQUEST,
            content={"message": "Invalid data URL"},
        )

    mime = match.group(1).lower()
    if not mime.startswith("image/"):
        return JSONResponse(
            status_code=status.HTTP_400_BAD_REQUEST,
            content={"message": "Only image uploads are supported"},
        )

    try:
        buf = base64.b64decode(match.group(2), validate=False)
    except Exception:
        return JSONResponse(
            status_code=status.HTTP_400_BAD_REQUEST,
            content={"message": "Invalid base64 payload"},
        )

    if len(buf) == 0:
        return JSONResponse(
            status_code=status.HTTP_400_BAD_REQUEST,
            content={"message": "Empty file"},
        )
    if len(buf) > MAX_BYTES:
        return JSONResponse(
            status_code=status.HTTP_413_REQUEST_ENTITY_TOO_LARGE,
            content={"message": "Image too large (max 12 MB)"},
        )

    fallback_ext = (mime.split("/", 1)[1] if "/" in mime else "bin") or "bin"
    fallback_ext = EXT_CLEAN_RE.sub("", fallback_ext)[:6] or "img"

    safe = SAFE_FILENAME_RE.sub("_", filename)[:60] or "image"
    has_ext = bool(HAS_EXT_RE.search(safe))

    stamp = secrets.token_hex(8)
    final_name = f"{stamp}-{safe}" if has_ext else f"{stamp}-{safe}.{fallback_ext}"

    UPLOADS_DIR.mkdir(parents=True, exist_ok=True)
    target = UPLOADS_DIR / final_name
    target.write_bytes(buf)

    return {"url": f"/api/uploads/{final_name}", "bytes": len(buf)}
